### PR TITLE
DAOS-16111 rebuild: uniform identifier in logs part 1

### PR DESCRIPTION
**Pull request by kccain**
To the extent possible in the rebuild code execution flow, when rebuild emits log messages, include a uniform rebuild operation identifier in those messages. This covers activities across all pool storage engines (including the pool service leader), system and per-target threads/xstreams, and dynamically spawned user-level threads.

The motivation is to enable some amount of automated searching through logfiles for all (or specific) rebuilds that occurred during execution, and speed up DAOS engineer analysis/interpretation of the logs.

The baseline format (defined in the DF_RB macro) is:
  "rb=" DF_UUID "/%u/%u/%s"
and corresponds to:
   <pool_uuid>/<rebuild_ver>/<rebuild_gen>/<opstring>

A verbose format (defined in the DF_RBF macro) adds the following (for <leader_rank>/<term>)
" ld=%u/" DF_U64

Various DP_RB_* and DP_RBF_* macros are defined to specify the arguments to go with the DF_RB and DF_RBF formats, given some common rebuild implementation structures such as:
  struct rebuild_global_pool_tracker
  struct rebuild_tgt_pool_tracker
  struct rebuild_scan_in (REBUILD_OBJECTS_SCAN RPC input)
  struct migrate_query_arg

This initial patch covers the pool service leader execution in functions (and those that they invoke) such as:
  rebuild_ults()
  rebuild_task_ult()
  rebuild_leader_start()
  rebuild_leader_status_check()
  rebuild_leader_status_notify().

And this patch covers "scan side" execution in all pool storage engines (including the leader), in functions such as:
  rebuild_tgt_scan_handler()
  rebuild_tgt_status_check_ult()
  ds_migrate_query_status()
  migrate_check_one()
  dss_rebuild_check_one()
  rebuild_scan_leader()
  rebuild_scanner()
  rebuild_objects_send_ult()
  rebuild_scan_done()

Features: rebuild
Allow-unstable-test: true
Skip-list: test_ec_degrade:DAOS-15843

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Comment by kccain** [comment]
@liuxuezhao @gnailzenh here is a draft patch for what I discussed today. Can you take a look and provide some early comments on the approach?

File: src/include/daos_srv/daos_engine.h
Comment (liuxuezhao) [inline src/include/daos_srv/daos_engine.h:696]: not a problem of this patch, the ds_migrate_xxx and ds_object_migrate_send declaration looks better to be in "daos_srv/object.h" rather in daos_engine.h

**Review by kccain — COMMENTED** [review]
addressed feedback in the latest push

File: src/include/daos_srv/rebuild.h
Comment (liuxuezhao) [reply src/include/daos_srv/rebuild.h]: thanks, D_LOG_FILE_APPEND_RANK can be used to append #rank to the log file name, is it convenient for you to use? Some times a failure happened, but when we enable the log mask it cannot be reproduce anymore, so we may need to consider to log as less as possible if it can prove enough information.

**Comment by kccain** [comment]
all per-PR tests passed. Some full_regression (weekly regresison tests) failed and are known issues (listed at the end)

But this one from Functional HW Large [testing]( https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14383/10/execution/node/651/log) that includes an engine assertion seems new.] I don't see a direct link between the changes in this PR and the aggregation engine [assertion and stacktrace](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14383/10/artifact/Functional%20Hardware%20Large/erasurecode/offline_rebuild_aggregation.py/stacktraces.wolf-113/'core.daos_io_5.1716943724.185489.stacktrace'/*view*/). @liuxuezhao do you think this patch could have caused it? Or possibly it is a rare issue not yet seen in the testing?
- erasurecode/offline_rebuild_aggregation.py [failure](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14383/10/testReport/junit/FTEST_erasurecode/EcodAggregationOffRebuild/Test_Hardware___Functional_Hardware_Large___2___erasurecode_offline_rebuild_aggregation_py_EcodAggregationOffRebuild_test_ec_offline_agg_during_rebuild_run_container_hosts_ior_client_processes_iorflags_objectclass_pool_server_config_engines_0_storage_0_1_setup_70cf/)


Functional HW Large [testing]( https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14383/10/execution/node/651/log) with known existing failures:
- ior/hard_rebuild.py failure is an instance of [DAOS-15863](https://daosio.atlassian.net/browse/DAOS-15863)

Functional HW Large MD on SSD [testing ](https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14383/10/execution/node/667/log) with known existing failures:
- erasurecode/online_rebuild.py cases [2](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14383/10/testReport/junit/FTEST_erasurecode/EcodOnlineRebuild/Test_Hardware___Functional_Hardware_Large_MD_on_SSD___2___erasurecode_online_rebuild_py_EcodOnlineRebuild_test_ec_online_rebuild_run_agent_config_container_daos_destroy_hosts_servers_6_server_ior_client_processes_iorflags_objectclass_sizes_Partial_Striped_pool_server_config_control_metadata_engines_0_storage_0_1_2_setup_ad15/) and [4](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-14383/10/testReport/junit/FTEST_erasurecode/EcodOnlineRebuild/Test_Hardware___Functional_Hardware_Large_MD_on_SSD___4___erasurecode_online_rebuild_py_EcodOnlineRebuild_test_ec_online_rebuild_run_agent_config_container_daos_destroy_hosts_servers_8_server_ior_client_processes_iorflags_objectclass_sizes_Partial_Striped_pool_server_config_control_metadata_engines_0_storage_0_1_2_setup_f863/) failures are also an instance of [DAOS-15863](https://daosio.atlassian.net/browse/DAOS-15863)

Functional HW Medium MD on SSD [testing](https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14383/10/execution/node/505/log) failures:
- deployment/server_rank_failure.py cases 1 and 2 failures are instances of [DAOS-15809](https://daosio.atlassian.net/browse/DAOS-15809)

File: src/object/srv_obj_migrate.c
Comment (liuxuezhao) [inline src/object/srv_obj_migrate.c:4018]: ditto, need not specially log unk

**Comment by liuxuezhao** [comment]
"@liuxuezhao do you think this patch could have caused it? Or possibly it is a rare issue not yet seen in the testing?"
I think that assertion in vos agg code path should be unrelated with this PR, I created https://daosio.atlassian.net/browse/DAOS-15941 for it.

File: src/include/daos_srv/rebuild.h
Comment (kccain) [reply src/include/daos_srv/rebuild.h:43]: That is correct. DF_RBF was intented to include leader_rank as the final value, and I missed updating the comment.

If I remove leader_rank here, then the remaining information would be the PS term. Shall we keep the term in the DF_RBF rebuild log messages? If not, then there is no need for DF_RBF at all.

For now, until I see your response, I will update the patch so DF_RBF will not include the leader_rank, and keep the term value.

Comment (liuxuezhao) [reply src/include/daos_srv/rebuild.h:43]: I think need not log the leader rank/term in every log entry, we may only log that info when rebuild code detect leader switched, for example in rebuild_tgt_scan_handler() it found "rpt->rt_leader_rank != rsi->rsi_master_rank".
So seems that can keep DF_RBF and still with leader_rank and leader_term, but only use it in special place with leader switch, rather than add it to all/most log entries. What do you think? thanks for considering it.

Comment (kccain) [reply src/include/daos_srv/rebuild.h]: Yes, let's start with this approach, minimizing its use, and see if we may find other usages for DF_RBF later. I've updated DF_RBF and usage sites in the latest push.

File: src/object/srv_obj_migrate.c
Comment (kccain) [reply src/object/srv_obj_migrate.c]: Fixed in the latest push.

**Comment by kccain** [comment]
> Test stage Functional on EL 8.8 completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14383/13/execution/node/1153/log

looks like some CI infrastructure issue "CommandFailure("ERROR: Copying yaml configuration file to wolf-231vm[1,8]: Timeout detected running 'clush -S -v -w wolf-231vm[1,8] /usr/bin/mkdir -p /etc/daos' with a 15s timeout",)"

Kicking off the testing again from the Test stage ...

File: src/rebuild/scan.c
Comment (liuxuezhao) [inline src/rebuild/scan.c:327]: this one looks can be D_ERROR

File: src/rebuild/srv.c
Comment (liuxuezhao) [inline src/rebuild/srv.c:1425]: here the "task->dst_rebuild_op" possibly be different with rgt->rgt_opc.
For example a REBUILD failed and here schedule a FAIL_RECLAIM task.
So better to log the retry opc same as original code, to avoid confuse.
same for several other places in this function.

Comment (kccain) [reply src/rebuild/srv.c:1425]: In this function I'll restore the logging to have the task->dst_rebuild_op and remove DF_RB.

Maybe in a future patch, if it is functionally correct, I could use a format like this to get part of DF_RB information while keeping the task information too:
DF_UUID "%u/%u opc %u/%u ...",
DP_UUID(), rgt->rgt_rebuild_ver, rgt->rgt_rebuild_gen, task->dst_rebuild_op, task->dst_rebuild_gen

Doing that can make it easier to find these slightly different log messages with the other logs that use DF_RB

File: src/rebuild/scan.c
Comment (kccain) [reply src/rebuild/scan.c:327]: will change to an error log message

**Review by liuxuezhao — APPROVED** [review]
LGTM, thanks @kccain

**Comment by kccain** [comment]
CI testing is blocked at the moment by https://daosio.atlassian.net/browse/SRE-2233
This PR has passed Features: rebuild testing in build 10 before, so I will move it from draft -> review-able status now and we can get a second reviewer in parallel to CI testing, once that's ready to resume.

File: src/tests/ftest/erasurecode/offline_rebuild_single.yaml
Comment (daltonbohning) [inline src/tests/ftest/erasurecode/offline_rebuild_single.yaml:15]: Is this an optimization or necessary for these tests? In other words, what would happen with a similar workflow in production?

Comment (kccain) [reply src/tests/ftest/erasurecode/offline_rebuild_single.yaml:15]: I made these .yaml file changes to address intermittent overall test timeout failures in some in Features: rebuild weekly regression tests. As seen by libdaos programs (like the 'daos' command) getting cart proto query RPC timeouts in the libdaos bootstrapping code. When it randomly chooses a stopped engine as the destination of that RPC. But I really wanted those tests to be run with my engine C code changes, to make sure they did not accidentally introduce a regression.

It is a similar change to what was done for other tests https://github.com/daos-stack/daos/pull/14086
A longer term solution would be to improve the bootstrapping code logic in libdaos to shorten timeouts and try alternate engines. Some discussion in https://daosio.atlassian.net/browse/DAOS-15440 and several other tickets and discussion channels IIRC.

Comment (daltonbohning) [reply src/tests/ftest/erasurecode/offline_rebuild_single.yaml:15]: Gotcha. Since we're tracking an improvement, no issue from me

Comment (jolivier23) [reply src/tests/ftest/erasurecode/offline_rebuild_single.yaml:15]: I thought @mjmac was using this to workaround an issue where slurm replaces the user/group stuff.  You may want to sync with him.

**Comment by kccain** [comment]
> Test stage Functional Hardware Medium completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14383/17/execution/node/1386/log

The single test failure seen in this testing is an instance of existing known issue https://daosio.atlassian.net/browse/DAOS-14585

**Comment by kccain** [comment]
> Test stage Functional Hardware Medium UCX Provider completed with status FAILURE. https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-14383/17/execution/node/1524/log

The single test failure seen in this testing (daos_test/suite.py:DaosCoreTest.test_daos_single_rdg_tx) is an instance of existing known issue https://daosio.atlassian.net/browse/DAOS-14982

**Comment by kccain** [comment]
Features testing done/successful in build 10.
PR testing done/successful in build 17

Patch requires one more peer review

**Comment by kccain** [comment]
While waiting for reviews, I am continuing to investigate if the weekly regression test failure (recx2ext() engine assertion) seen in build 10 is always associated with this patch, or if it can be seen in master. That work is being documented in https://daosio.atlassian.net/issues/DAOS-15941

For now I have only seen it with this patch.

**Review by wangshilong — APPROVED** [review]

File: src/rebuild/srv.c
Comment (kccain) [inline src/rebuild/srv.c:901]: a mistake here? Should the DL_ERROR(rc, ...) be placed after the assignment rc = ret?

or DL_ERROR(ret, ...) here.

Note for the next push, which I didn't think would be needed. But maybe due to this issue I'll need to merge with latest master and push for testing.

Comment (kccain) [reply src/rebuild/srv.c:901]: Instead, I'll make the fix in the part 2 patch https://github.com/daos-stack/daos/pull/14547

**Comment by daltonbohning** [comment]
@kccain Is there a ticket to track this work? IMO it's better for changes like this to have a ticket for tracking purposes

**Comment by kccain** [comment]
> @kccain Is there a ticket to track this work? IMO it's better for changes like this to have a ticket for tracking purposes

No ticket yet. I will create one, especially useful also if request for merge to a release branch should be desired.

**Review by daltonbohning — CHANGES_REQUESTED** [review]
-1 just until we get a ticket number to put in the commit

**Comment by kccain** [comment]
> -1 just until we get a ticket number to put in the commit

https://daosio.atlassian.net/browse/DAOS-16111 has been filed

I've edited the title of this PR in the github web UI to include this ticket number rather than the generic DAOS-623.

**Pull request merged**
